File: spicedb-rust/tests/raw_escape_hatch_test.rs
//! The escape hatch reaches a real stub and makes a real call.
//!
//! `SpiceDBClient::raw_proto()` exists so a request the idiomatic surface
//! cannot express has a workaround short of forking the crate -- root
//! DESIGN.md, "What NOT To Do", permits exactly this as "clearly marked
//! secondary API". Asserting that the accessor compiles, or that the returned
//! value has a `permissions` field, would prove none of that. What matters is
//! whether a caller can drive a generated tonic client through it and get an
//! answer out of a real server, with this client's bearer token attached.
//!
//! So these tests stand up a real in-process tonic server (the shared
//! `support` harness) and drive `CheckPermission` -- the single-check RPC the
//! idiomatic client never calls, since every check it makes goes through
//! `CheckBulkPermissions`. That makes it a genuine gap rather than a
//! contrived one.

mod support;

use spicedb::client::SpiceDBClient;
use spicedb::spicedb_proto::authzed::api::v1 as proto;
use support::{spawn_permissions_server, MockPermissionsService};

fn check_request() -> proto::CheckPermissionRequest {
    proto::CheckPermissionRequest {
        consistency: Some(proto::Consistency {
            requirement: Some(proto::consistency::Requirement::FullyConsistent(true)),
        }),
        resource: Some(proto::ObjectReference {
            object_type: "document".to_string(),
            object_id: "readme".to_string(),
        }),
        permission: "view".to_string(),
        subject: Some(proto::SubjectReference {
            object: Some(proto::ObjectReference {
                object_type: "user".to_string(),
                object_id: "jimmy".to_string(),
            }),
            optional_relation: String::new(),
        }),
        context: None,
        with_tracing: false,
    }
}

#[tokio::test]
async fn raw_proto_drives_a_real_generated_client_against_a_real_server() {
    let mock = MockPermissionsService::new();
    let requests = mock.check_permission_requests();
    let authorizations = mock.check_permission_authorizations();
    let addr = spawn_permissions_server(mock).await;

    let client = SpiceDBClient::new_plaintext(addr.to_string(), "test-token")
        .await
        .expect("client should connect to the mock server");

    // Clone, because the generated clients take `&mut self`. A tonic client
    // clone shares the same channel -- it does not open a second connection.
    let mut permissions = client.raw_proto().permissions.clone();
    let response = permissions
        .check_permission(check_request())
        .await
        .expect("raw CheckPermission should succeed")
        .into_inner();

    assert_eq!(
        response.permissionship,
        proto::check_permission_response::Permissionship::HasPermission as i32,
        "the raw call must return the server's real answer"
    );
    assert_eq!(
        response.checked_at.expect("a revision").token,
        "rev-raw",
        "the response must be the one this server produced"
    );

    let received = requests.lock().unwrap();
    assert_eq!(received.len(), 1, "the RPC must have reached the server");
    assert_eq!(received[0].permission, "view");

    // The bearer token rides this client's interceptor, so a raw call is
    // authenticated exactly as an idiomatic one is -- nothing extra to pass.
    assert_eq!(
        authorizations.lock().unwrap().as_slice(),
        &["Bearer test-token".to_string()],
    );
}

#[tokio::test]
async fn raw_proto_shares_the_connection_the_idiomatic_methods_use() {
    let mock = MockPermissionsService::new();
    mock.push_check_bulk_permissions_response(proto::CheckBulkPermissionsResponse {
        pairs: vec![proto::CheckBulkPermissionsPair {
            request: None,
            response: Some(proto::check_bulk_permissions_pair::Response::Item(
                proto::CheckBulkPermissionsResponseItem {
                    permissionship: proto::check_permission_response::Permissionship::HasPermission
                        as i32,
                    partial_caveat_info: None,
                    debug_trace: None,
                },
            )),
        }],
        checked_at: Some(proto::ZedToken {
            token: "rev-bulk".to_string(),
        }),
    });
    let bulk_calls = mock.check_bulk_permissions_calls();
    let raw_calls = mock.check_permission_requests();
    let addr = spawn_permissions_server(mock).await;

    let client = SpiceDBClient::new_plaintext(addr.to_string(), "test-token")
        .await
        .expect("client should connect to the mock server");

    // An idiomatic call and a raw one, in that order, over one client.
    let rel = spicedb::types::Relationship::new("document", "readme", "view", "user", "jimmy", "")
        .expect("valid relationship");
    let result = client
        .check_permission(&spicedb::consistency::full(), "view", &rel)
        .await
        .expect("idiomatic check should succeed");
    assert!(result.has_permission());

    let mut permissions = client.raw_proto().permissions.clone();
    permissions
        .check_permission(check_request())
        .await
        .expect("raw CheckPermission should succeed");

    assert_eq!(
        bulk_calls.load(std::sync::atomic::Ordering::SeqCst),
        1,
        "the idiomatic call must have gone through CheckBulkPermissions"
    );
    assert_eq!(
        raw_calls.lock().unwrap().len(),
        1,
        "the raw call must have gone through the single-check RPC the idiomatic API never uses"
    );
}

/// The hatch must never grow into a way to build a connection.
///
/// Root DESIGN.md, "RULE: Credentials over insecure transport require an
/// explicit opt-in", is enforced in `SpiceDBClientBuilder::build`, on the
/// single path that creates a channel. Handing back an already-built client
/// cannot bypass that; accepting an endpoint, token, or transport setting here
/// would -- it would be a second construction path with no guard on it.
#[tokio::test]
async fn raw_proto_is_an_accessor_not_a_second_construction_path() {
    // `raw_proto` takes only `&self`: this would not compile if it took
    // connection parameters, which is the shape that makes a bypass possible.
    let accessor: fn(&SpiceDBClient) -> &spicedb::spicedb_proto::SpiceDBProtoClient =
        SpiceDBClient::raw_proto;
    let _ = accessor;

    // And the guard still refuses what it always did, so nothing has moved.
    let err = match SpiceDBClient::builder("evil.example.com:50051", "token")
        .plaintext()
        .build()
        .await
    {
        Ok(_) => panic!("plaintext to a non-loopback host must still be refused"),
        Err(err) => err,
    };
    assert!(
        err.to_string()
            .contains("allow_insecure_remote_credentials"),
        "unexpected error: {err}"
    );
}

File: spicedb-java/lib/src/test/java/com/authzed/spicedb/RawEscapeHatchTest.java
package com.authzed.spicedb;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import build.buf.gen.authzed.api.v1.CheckPermissionRequest;
import build.buf.gen.authzed.api.v1.CheckPermissionResponse;
import build.buf.gen.authzed.api.v1.Consistency;
import build.buf.gen.authzed.api.v1.ObjectReference;
import build.buf.gen.authzed.api.v1.PermissionsServiceGrpc;
import build.buf.gen.authzed.api.v1.SubjectReference;
import build.buf.gen.authzed.api.v1.ZedToken;
import com.authzed.spicedb.errors.InvalidArgumentException;
import io.grpc.Metadata;
import io.grpc.Server;
import io.grpc.ServerCall;
import io.grpc.ServerCallHandler;
import io.grpc.ServerInterceptor;
import io.grpc.inprocess.InProcessChannelBuilder;
import io.grpc.inprocess.InProcessServerBuilder;
import io.grpc.stub.StreamObserver;
import java.io.IOException;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import org.junit.jupiter.api.Test;

/**
 * The escape hatch, {@link SpiceDBClient#rawChannel()}, exists so a request the idiomatic surface
 * cannot express has a workaround short of forking the client. Asserting the accessor returns
 * something non-null would prove none of that. What matters is whether a caller can build a
 * generated stub on it and get an answer out of a real server, with this client's bearer token
 * attached -- so these tests run a real (in-process) gRPC server and assert the {@code
 * authorization} metadata the server actually received.
 *
 * <p>The RPC driven here is {@code CheckPermission}, the single-check call the idiomatic client
 * never makes ({@link SpiceDBClient#checkPermission} routes every check through {@code
 * CheckBulkPermissions}), so the gap is genuine rather than contrived.
 */
class RawEscapeHatchTest {

  private static final Metadata.Key<String> AUTHORIZATION_KEY =
      Metadata.Key.of("authorization", Metadata.ASCII_STRING_MARSHALLER);

  /**
   * A real in-process server serving both the single-check and bulk-check RPCs, recording the
   * {@code authorization} metadata of every call in arrival order via a server-level interceptor.
   */
  private static final class RecordingServer implements AutoCloseable {
    final Server server;
    final String name;
    final List<String> capturedAuth = new CopyOnWriteArrayList<>();

    RecordingServer() throws IOException {
      this.name = InProcessServerBuilder.generateName();
      ServerInterceptor interceptor =
          new ServerInterceptor() {
            @Override
            public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
                ServerCall<ReqT, RespT> call,
                Metadata headers,
                ServerCallHandler<ReqT, RespT> next) {
              capturedAuth.add(headers.get(AUTHORIZATION_KEY));
              return next.startCall(call, headers);
            }
          };
      this.server =
          InProcessServerBuilder.forName(name)
              .directExecutor()
              .addService(
                  new PermissionsServiceGrpc.PermissionsServiceImplBase() {
                    @Override
                    public void checkPermission(
                        CheckPermissionRequest request,
                        StreamObserver<CheckPermissionResponse> observer) {
                      observer.onNext(
                          CheckPermissionResponse.newBuilder()
                              .setPermissionship(
                                  CheckPermissionResponse.Permissionship
                                      .PERMISSIONSHIP_HAS_PERMISSION)
                              .setCheckedAt(ZedToken.newBuilder().setToken("rev-raw").build())
                              .build());
                      observer.onCompleted();
                    }

                    @Override
                    public void checkBulkPermissions(
                        build.buf.gen.authzed.api.v1.CheckBulkPermissionsRequest request,
                        StreamObserver<build.buf.gen.authzed.api.v1.CheckBulkPermissionsResponse>
                            observer) {
                      var response =
                          build.buf.gen.authzed.api.v1.CheckBulkPermissionsResponse.newBuilder();
                      for (int i = 0; i < request.getItemsCount(); i++) {
                        response.addPairs(
                            build.buf.gen.authzed.api.v1.CheckBulkPermissionsPair.newBuilder()
                                .setItem(
                                    build.buf.gen.authzed.api.v1.CheckBulkPermissionsResponseItem
                                        .newBuilder()
                                        .setPermissionship(
                                            CheckPermissionResponse.Permissionship
                                                .PERMISSIONSHIP_HAS_PERMISSION)
                                        .build())
                                .build());
                      }
                      observer.onNext(response.build());
                      observer.onCompleted();
                    }
                  })
              .intercept(interceptor)
              .build()
              .start();
    }

    InProcessChannelBuilder channelBuilder() {
      return InProcessChannelBuilder.forName(name).directExecutor();
    }

    @Override
    public void close() {
      server.shutdownNow();
    }
  }

  private static CheckPermissionRequest request() {
    return CheckPermissionRequest.newBuilder()
        .setConsistency(Consistency.newBuilder().setFullyConsistent(true).build())
        .setResource(
            ObjectReference.newBuilder().setObjectType("document").setObjectId("readme").build())
        .setPermission("view")
        .setSubject(
            SubjectReference.newBuilder()
                .setObject(
                    ObjectReference.newBuilder().setObjectType("user").setObjectId("jimmy").build())
                .build())
        .build();
  }

  private static SpiceDBClient clientFor(RecordingServer server) {
    return SpiceDBClient.create(
        "localhost:50051",
        "test-token",
        SpiceDBClient.DEFAULT_TIMEOUT,
        server.channelBuilder(),
        SpiceDBClient.withInsecure());
  }

  @Test
  void rawChannelDrivesARealStubAgainstARealServer() throws Exception {
    try (RecordingServer server = new RecordingServer()) {
      try (SpiceDBClient client = clientFor(server)) {
        CheckPermissionResponse response =
            PermissionsServiceGrpc.newBlockingStub(client.rawChannel()).checkPermission(request());

        assertEquals(
            CheckPermissionResponse.Permissionship.PERMISSIONSHIP_HAS_PERMISSION,
            response.getPermissionship());
        assertEquals("rev-raw", response.getCheckedAt().getToken());
      }

      // The bearer token rides the channel this client hands out, so a raw call is
      // authenticated exactly as an idiomatic one is -- nothing extra to attach.
      assertEquals(List.of("Bearer test-token"), server.capturedAuth);
    }
  }

  @Test
  void rawChannelSharesTheConnectionTheIdiomaticMethodsUse() throws Exception {
    try (RecordingServer server = new RecordingServer()) {
      try (SpiceDBClient client = clientFor(server)) {
        CheckResult result =
            client.checkPermission(
                com.authzed.spicedb.Consistency.full(),
                "view",
                Relationship.of("document", "readme", "view", "user", "jimmy"));
        assertTrue(result.hasPermission());

        PermissionsServiceGrpc.newBlockingStub(client.rawChannel()).checkPermission(request());
      }

      // One idiomatic call (via CheckBulkPermissions) and one raw call (via the single-check
      // RPC), both authenticated, both on this client's own connection.
      assertEquals(List.of("Bearer test-token", "Bearer test-token"), server.capturedAuth);
    }
  }

  /**
   * The hatch must never grow into a way to build a connection. Root DESIGN.md, "RULE: Credentials
   * over insecure transport require an explicit opt-in", is enforced in {@code create}, on the
   * single path that builds a channel. Handing back an already-built channel cannot bypass that;
   * accepting an endpoint, key, or transport setting would.
   */
  @Test
  void rawChannelIsAnAccessorNotASecondConstructionPath() throws Exception {
    assertEquals(
        0,
        SpiceDBClient.class.getMethod("rawChannel").getParameterCount(),
        "rawChannel must take no arguments");

    // And the guard still refuses what it always did.
    InvalidArgumentException ex =
        assertThrows(
            InvalidArgumentException.class,
            () ->
                SpiceDBClient.create(
                    "evil.example.com:1234", "test-token", SpiceDBClient.withInsecure()));
    assertTrue(ex.getMessage().contains("allowInsecureRemoteCredentials"), ex.getMessage());
  }
}
